lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetMetadataUtils.java

```diff
@@ -0,0 +1,229 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.writer;
+
+import org.apache.parquet.column.statistics.BinaryStatistics;
+import org.apache.parquet.format.Statistics;
+import org.apache.parquet.internal.column.columnindex.BinaryTruncator;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.schema.ColumnOrder;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+// Based on org.apache.parquet.format.converter.ParquetMetadataConverter
+public class ParquetMetadataUtils
+{
+    private static final long MAX_STATS_SIZE = 4096; // limit stats to 4k
+
+    private ParquetMetadataUtils() {}
+
+    public static <T extends Comparable<T>> Statistics toParquetStatistics(org.apache.parquet.column.statistics.Statistics<T> stats, int truncateLength)
+    {
+        Statistics formatStats = new Statistics();
+        // Don't write stats larger than the max size rather than truncating. The
+        // rationale is that some engines may use the minimum value in the page as
+        // the true minimum for aggregations and there is no way to mark that a
+        // value has been truncated and is a lower bound and not in the page.
+        if (!stats.isEmpty() && withinLimit(stats, truncateLength)) {
+            formatStats.setNull_count(stats.getNumNulls());
+            if (stats.hasNonNullValue()) {
+                byte[] min;
+                byte[] max;
+
+                if (stats instanceof BinaryStatistics && truncateLength != Integer.MAX_VALUE) {
+                    BinaryTruncator truncator = BinaryTruncator.getTruncator(stats.type());
+                    min = truncateMin(truncator, truncateLength, stats.getMinBytes());
+                    max = truncateMax(truncator, truncateLength, stats.getMaxBytes());
+                }
+                else {
+                    min = stats.getMinBytes();
+                    max = stats.getMaxBytes();
+                }
+                // Fill the former min-max statistics only if the comparison logic is
+                // signed so the logic of V1 and V2 stats are the same (which is
+                // trivially true for equal min-max values)
+                if (sortOrder(stats.type()) == SortOrder.SIGNED || Arrays.equals(min, max)) {
+                    formatStats.setMin(min);
+                    formatStats.setMax(max);
+                }
+
+                if (isMinMaxStatsSupported(stats.type()) || Arrays.equals(min, max)) {
+                    formatStats.setMin_value(min);
+                    formatStats.setMax_value(max);
+                }
+            }
+        }
+        return formatStats;
+    }
+
+    private static boolean withinLimit(org.apache.parquet.column.statistics.Statistics stats, int truncateLength)
```

Comment:
fix raw usage here too (`<?>`)

Reply:
This is deleted in the next commit, it's here just for ease of review